ERCS/erc-7629.md

```diff
@@ -0,0 +1,322 @@
+---
+eip: 7629
+title: Unified Token
+description: Unifying ERC-721 and ERC-20 for seamless asset conversion.
+author: 0xZeus1111 (@0xZeus1111), Nvuwa (@Nvuwa)
+discussions-to: https://ethereum-magicians.org/t/erc-7629-unified-token/18793
+status: Draft
+type: Standards Track
+category: ERC
+created: 2024-02-18
+requires: 20, 165, 721
+---
+
+
+## Abstract
+
+[ERC-7629](./eip-7629.md), known as the Unified Token, introduces a comprehensive protocol unifying the characteristics of [ERC-721](./eip-721.md) and [ERC-20](./eip-20.md) tokens within the Ethereum ecosystem. This standard seamlessly integrates the liquidity features of [ERC-20](./eip-20.md) with the non-fungible nature of [ERC-721](./eip-721.md), enabling frictionless conversion between these asset types. [ERC-7629](./eip-7629.md) offers a multifunctional solution, providing developers and users with the flexibility to leverage both liquidity and non-fungibility in a unified token framework.
+
+
+## Motivation
+
+The motivation behind [ERC-7629](./eip-7629.md) stems from the inherent need within the blockchain community for assets that possess both the liquidity of [ERC-20](./eip-20.md) tokens and the non-fungibility of [ERC-721](./eip-721.md) tokens. Current standards present a dichotomy, necessitating users to choose between these features. [ERC-7629](./eip-7629.md) addresses this limitation by providing a unified token standard, empowering users to seamlessly transition between [ERC-20](./eip-20.md) and [ERC-721](./eip-721.md) characteristics, catering to diverse blockchain applications.
+
+## Specification
+
+- [ERC-7629](./eip-7629.md) standardizes a token contract that encompasses features from both [ERC-20](./eip-20.md) and [ERC-721](./eip-721.md).
+- The token contract supports state transitions between [ERC-20](./eip-20.md) and [ERC-721](./eip-721.md) modes, ensuring smooth conversion and utilization of both liquidity and non-fungibility.
+- Defines essential functions and events to support token interactions, conversions, and queries.
+- Implements a low gas consumption [ERC-20](./eip-20.md) mode to maintain gas efficiency comparable to typical [ERC-20](./eip-20.md) token transfers.
+
+Compliant contracts MUST implement the following Solidity interface:
+
+```solidity
+
+pragma solidity ^0.8.0;
+/**
+ * @title ERC-7629 Unify Token Interface
+ * @dev This interface defines the ERC-7629 Unify Token, which unifies ERC-721 and ERC-20 assets.
+ */
+interface IERC7629 is IERC165{
+    // ERC-20 Transfer event
+    event ERC20Transfer(
+        address indexed from,
+        address indexed to,
+        uint256 amount
+    );
+
+    // ERC-721 Transfer event
+    event ERC721Transfer(
+        address indexed from,
+        address indexed to,
+        uint256 indexed tokenId
+    );
+
+    // Approval event for ERC-20 and ERC-721
+    event Approval(
+        address indexed owner,
+        address indexed approved,
+        uint256 indexed tokenId
+    );
```

Comment:
I've reviewed the proposed event signature for the unified Approval event intended to serve both ERC-20 and ERC-721 tokens, while feeling concerned about the potential ambiguity the current design might introduce. Specifically, it's unclear how to differentiate whether the third parameter, tokenId, represents an ERC-721 token's unique identifier or an ERC-20 token's value. In ERC-20 contexts, this parameter would typically denote a quantity, whereas in ERC-721 contexts, it uniquely identifies a token.